


Import libraries used in the analysis and set colours for each species group

```{r, message=FALSE}

setwd('C://Users/ywang120/Dropbox/iMutations/')

library('readxl')
library('data.table')
library('dplyr')
library('tidyr')

library('ggplot2')
library("cowplot")
library("RColorBrewer")
library("grid")
library("gtable")
library('ggpubr')
library("ggrepel")
library("ggtree")

library("MCMCglmm")
library("Hmisc")


mycol <- brewer.pal(12, 'Set3')

human_col       = mycol[4]
primates_col    = mycol[1]
mammals_col     = mycol[2]
birds_col       = mycol[3]
fish_col        = mycol[5]
arthropods_col  = mycol[6]
worm_col        = mycol[7]
plants_col      = mycol[8] 
fungus_col      = mycol[10]
unicellular_col = mycol[11]
```


Define some functions that will be used later. `inferEvents()` is to approximate the 
number of  mutation events based on the reported mutation rate and it's
intervals assuming a binomial distribution.


```{r}
###############################################
## FUNCTIONS

# infer effective number of events based on interval assuming binomial distribution
inferEvents <- function(um, u_lower, u_upper){
    delta = 100
    i = 1
    while (TRUE){
        tt <- binconf(i,i/um)
        t_lower <- tt[2]
        t_upper <- tt[3]
        low_diff <- (t_lower - u_lower)/u_lower
        up_diff <- (t_upper - u_upper)/u_upper
        tmp <- abs(low_diff) + abs(up_diff)
        if(tmp > delta){
            break
        }else{
            delta <- tmp
            i = i + 1
        }
    }
    return(i-1)
}

# get lower CI
inferLowerCI <- function(x,n){
    tt <- binconf(x,n)
    return(tt[2])
}

# get upper CI
inferUpperCI <- function(x,n){
    tt <- binconf(x,n)
    return(tt[3])   
}


```


The imported data set contains some studies using phenotype approach or phylogeny 
approach to infer mutation rate, which is not the scope of this study. We only
selected studies using mutation accumulation approach `MA` or parent-offspring
approach `PO`.


```{r}
bb <- read_excel('mutation_rate_literature_updating.xlsx', sheet='raw') %>% 
    filter(Method %in% c('MA','PO')) %>%
    select(-c('range_low','range_up','SampleSize','Generations','Sequencing',
              'Tools', 'Depth', 'Citation2', 'Link', 'Populations')) %>%
    as.data.table()
str(bb)
```


> PID: the ID of publication
> Date: the date of publication
> Group: the species group
> Name: the common names for species
> Species: species
> SID: sample ID, a study may contain multiple samples/populations. We use SID to
distinguish them.
> TYPE: mutation type - snp, indel, deletion, insertion
> Unit: the unit of mutation rate
> u_mean: reported mutation rate
> u_lower: reported lower end of 95%CI of mutation rate
> u_upper: reported upper end of 95%CI of mutation rate
> SE: reported standard error for mutation rate
> Events: the number of mutation events
> Callable: the number of callable sites used to estimate mutation rate
> Method: PO or MA
> CITATION: publication citations





# Part One

Summary of searched literature, including the number of studies by year, and 
the proportion of species identified in literature.



```{r}
bb[,PID] %>% unique() %>% length() # Number of studies
bb$Species %>% unique() %>% length() # Mumber of species

```

Summary the number of studies by year: bar chart

```{r}
# Number of studies by year
b_year <- bb %>% select(c('PID', 'Method')) %>% unique() %>%
    tidyr::extract(PID, 'Year', '(....).*') %>%
    group_by(Year, Method) %>% count() %>%
    mutate(Year = as.integer(Year))

b_year$Method <- factor(b_year$Method, levels=c('MA', 'PO'))


# bar plot for number of studies by year
P_bar <- ggplot(b_year, aes(Year, n, fill=Method)) +
    geom_bar(stat='identity') +
    scale_y_continuous(name='Number of Studies',expand = c(0,0), limits = c(0,20)) +
    scale_x_continuous(breaks = seq(min(b_year$Year), max(b_year$Year), by = 1)) +
    scale_fill_manual(name='', values = c('MA'='#009999','PO'='darkorange3')) +
    theme_classic() +
    guides(fill=guide_legend(nrow=2)) + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.4),
          #panel.grid.major = element_line(),
          axis.title = element_text(size=15),
          axis.text = element_text(size=13),
          legend.text = element_text(size=16),
          legend.position = c(0.2,0.8),
          legend.key.size = unit(2, 'lines'),
          plot.margin = margin(3, 0.4, 3, 0.4, "cm"))

```


Summary the porportion of species: pir chart

```{r}
#####
# Number of studies by species
b_species <- bb %>% select(c('PID', 'Group')) %>% unique() %>%
    group_by(Group) %>% count()

b_species$Group <- factor(b_species$Group, 
                          levels=c('human','primates','mammals','birds',
                                   'fish','arthropods','worm','plants',
                                   'fungus','unicellular'))

b_species <- b_species %>% arrange(desc(Group)) %>%
    mutate(prop = n/sum(b_species$n) * 100) %>% as.data.frame() %>%
    mutate(ypos = cumsum(prop) - 0.5*prop) %>%
    mutate(ll = sprintf('%s\n%.1f%%', Group, prop))


scol <- c('human'=human_col,'primates'=primates_col,'mammals'=mammals_col,
          'birds'=birds_col, 'fish'=fish_col,'arthropods'=arthropods_col,
          'worm'=worm_col,'plants'=plants_col,
          'fungus'=fungus_col,'unicellular'=unicellular_col)


# Pie chart for proportion of species
P_pie_species <- ggplot(b_species, aes(x='', y=prop, fill=Group)) +
    geom_bar(width = 1, stat='identity', color='white') +
    scale_fill_manual(values=scol) +
    coord_polar('y',start=0) +
    theme_void() +
    #geom_text(aes(y=ypos, label=Group), size=5)+ 
    theme(legend.position = 'none') +
    geom_label_repel(aes(y=ypos, label=ll), size=4.5, nudge_x = 0.3, 
                     min.segment.length=unit(6,'lines'),
                     color='black', box.padding = 0.5, fill='white')

```


Output figure

```{r}

p12 <- plot_grid(P_pie_species, P_bar, nrow=1, labels = c( '(A)', '(B)'), label_x = 0.1)

# ggsave('study_summary.jpg', p12, width=12, height=8, units = 'in', dpi = 300)
ggsave('study_summary.pdf', p12, width=12, height=8, units = 'in', dpi = 300)

##
# b1 <- bb %>% select(c('PID', 'Group', 'Method')) %>% unique()
# b1[Group!='human', Method] %>% table
# b_year %>% as.data.table %>% filter(Year>=2010) %>%
#     select(n) %>% sum()
```



# Part Two

Single nucleotide mutations `SNM`

We create a new column `PID2` using author and year, e.g., `Wang, 2022`; and then
select snp data (SNM); and only include studies measure mutation rate as per bp per generation;
and filter out studies using exome, or male or sperm.


```{r}

bb_snp <- bb %>% extract(CITATION, 'Author', '^(.*?),') %>%
    tidyr::unite('PID2','Author','Date', sep=',') %>%
    filter(TYPE=='snp') %>% 
    filter(Unit=='bp/generation') %>%
    mutate(u_mean = as.numeric(u_mean)) %>% 
    mutate(u_lower = as.numeric(u_lower)) %>%
    mutate(u_upper = as.numeric(u_upper)) %>%
    mutate(SE = as.numeric(SE)) %>%
    mutate(Events = as.integer(Events)) %>%
    filter(!Comments %like% 'exome') %>%
    filter(!Comments %like% 'male') %>% 
    filter(!Comments %like% 'sperm') %>% 
    select(c('PID','PID2','SID','u_mean','u_lower','u_upper','Events','Callable', 'SE',
             'Name','Species','Method', 'Group')) %>%
    as.data.table

# Calculate interval if SE and u_mean are provided, u +/- 1.96 SE ?

bb_snp[is.na(Events) & is.na(u_lower) & !is.na(SE) & !is.na(u_mean), 
       u_lower:=u_mean-1.96*SE]
bb_snp[is.na(Events) & is.na(u_upper) & !is.na(SE) & !is.na(u_mean), 
       u_upper:=u_mean+1.96*SE]

# Estimate Events based on mean and interval
bb_snp[is.na(Events) & !is.na(u_lower), 
       Events:=apply(.SD,1,function(x) inferEvents(x[1],x[2],x[3])), 
       .SDcols=c('u_mean','u_lower','u_upper')]

# Assume 1 Event if interval no available (only u_mean provided)
bb_snp[is.na(Events) & is.na(u_lower), Events:=1]  


bb_snp[, Callable:=Events/u_mean]
bb_snp$PID <- as.character(bb_snp$PID)
```


Retrieve phylogenetic tree from <https://www.treetime.org>. For some species, there are no
available tree on treetime. We refer to <https://www.onezoom.org/> and use the most
closet species (`Species2` ) as a proxy in order to get a full tree from treetime.



```{r}
# repair unresolved species name by treetime.org
# https://www.onezoom.org/
# Using nearest species if not found in treetime
bb_snp[,Species2:=Species]
bb_snp[Species=='Amphilophus', Species2:='Amphilophus labiatus']
bb_snp[Species=='Anopheles stephensi', Species2:='Anopheles gambiae']
bb_snp[Species=='Anopheles coluzzii', Species2:='Anopheles gambiae']
bb_snp[Species %in% c('Paramecium sexaurelia','Paramecium biaurelia'), 
       Species2:='Paramecium tetraurelia']
bb_snp[Species=='Chlamydomonas incerta', Species2:='Chlamydomonas reinhardtii']
bb_snp[Species=='Daphnia galeata', Species2:='Daphnia dubia']
bb_snp[Species %in% c('Ostreococcus mediterraneus', 'Bathycoccus prasinos'), 
       Species2:='Ostreococcus tauri']
bb_snp[Species=='Saccharomycodes ludwigii', Species2:='Hanseniaspora valbyensis']
# following automatically replaced with neighbouring species by treetime.org
bb_snp[Species=='Micromonas pusilla', Species2:='Micromonas']
bb_snp[Species=='Picochlorum costavermella', Species2:='Prototheca wickerhamii']
bb_snp[Species=='Rhodotorula toruloides', Species2:='Rhodotorula graminis']
bb_snp[Species=='Bombus terrestris', Species2:='Bombus']
bb_snp[Species=='Chironomus riparius', Species2:='Chironomus pallidivittatus']
bb_snp[Species=='Pristionchus pacificus', Species2:='Pristionchus']


# bb_unique_species <- bb_snp[,"Species2"] %>% unique()
# fwrite(bb_unique_species, 'filtered_unique_species.txt', col.names = FALSE)

## go to: https://timetree.org/, upload filtered_unique_species.txt to get a full tree.
## Manually make adjust branch length to avoid length of zero by changing length 0.1

```


Based on the above species, we get tree file from timetree: `filtered_unique_species.nwk`.

To produce Figure2, we need a phylogentic tree with common names, instead of
species. For example, we use 'fruit fly' as a common name, which include D. melanogaster,
D. simulans, etc. 

We prune the tree to keep only one species with the same common names in the tree,
and replace the species name with common name.




```{r}
# create phylogeny based on species common names
my_phylo <- read.tree('filtered_unique_species.nwk')


# rename node labels as '^n\\d+'
my_phylo$node.label <- 
    stringr::str_replace_all(my_phylo$node.label, "'", "") %>% 
    paste0('n', .)


# human is not species, change its Group to primates
bb_snp[Group=='human', Group:='primates']

# random choose one species in each group as a proxy
bb_snp_name <- bb_snp[,.SD[1,], by='Name', .SDcols=c('Species2')] 
bb_snp_name$Species2 <- stringr::str_replace_all(bb_snp_name$Species2, ' ', '_')

# prune the tree, keep one species for each group (common name) in the tree
my_phylo_simple <- ape::keep.tip(my_phylo, tip=bb_snp_name$Species2)
my_phylo_simple <- ladderize(my_phylo_simple)



# rename tip labels as common names
for(nn in my_phylo_simple$tip.label){
    nn2 <- bb_snp_name[Species2==nn, Name]
    my_phylo_simple$tip.label <- 
        stringr::str_replace_all(my_phylo_simple$tip.label, nn, nn2)
}

plot(my_phylo_simple)

```



With tree and the data, we fit an phylogentic model: `mu ~ (1|phylogeny)`


```{r}

bb_snp$Group <- factor(bb_snp$Group, 
                       levels=c('unicellular','primates', 'mammals',  'birds',  
                                'fish', 'worm', 'arthropods', 'fungus','plants'))

InverseTree_simple <- inverseA(my_phylo_simple,scale=TRUE)$Ainv
bb_snp$log_callable <- log(bb_snp$Callable)

vv <- diag(2) 
diag(vv) <- c(1e6, 1e-8)
my_prior <- list(B=list(mu=c(0,1), V=vv),
                 G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))

m1 <- MCMCglmm(Events ~ log_callable,
               random= ~ Name,
               data=bb_snp,
               ginverse=list(Name=InverseTree_simple),
               prior = my_prior,
               nitt=5001000, thin=1000, burnin=10000,
               family = "poisson",
               pr=TRUE, verbose = FALSE)


## we try model with (1|species) as m1.x
bb_snp$Name_copy <-bb_snp$Name
my_prior.x <- list(B=list(mu=c(0,1), V=vv),
                 G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100),
                        G2=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))

m1.x <- MCMCglmm(Events ~ log_callable,
               random= ~ Name + Name_copy,
               data=bb_snp,
               ginverse=list(Name=InverseTree_simple),
               prior = my_prior.x,
               nitt=5001000, thin=1000, burnin=10000,
               family = "poisson",
               pr=TRUE, verbose = FALSE)

# m1 and m1.x are quite similar


HPDinterval((m1.x$VCV[,'Name'] + m1.x$VCV[,'Name_copy'])/rowSums(m1.x$VCV))
mean((m1.x$VCV[,'Name'] + m1.x$VCV[,'Name_copy'])/rowSums(m1.x$VCV))


HPDinterval((m1.x$VCV[,'Name'])/rowSums(m1.x$VCV))
mean((m1.x$VCV[,'Name'])/rowSums(m1.x$VCV))



# create a table for prediction
bb_snp_pred <- data.table('Name'=c(my_phylo_simple$node.label, 
                                   my_phylo_simple$tip.label)) %>%
    filter(Name!='n') %>%
    mutate(Events=0) %>%
    mutate(log_callable=0)

bb_snp_pred[Name=='n9', log_callable:=1]

# prediction based on model m1
bb_snp_pred <- predict(m1, interval="confidence", newdata=bb_snp_pred, 
                       type='response', marginal = NULL) %>%
    cbind(bb_snp_pred,.)
```


Figure 2: Based on the prediction, we plot the forest tree for mutation rate. 
The plot will be further edited for aesthetic purpose.


```{r}
# Plot whole phylogenetic tree
plot(my_phylo_simple)

nodelabels(text = my_phylo_simple$node.label, cex=0.6)

## get root mean
anc0 <- bb_snp_pred[Name=='n148',4:6]
log_anc0 <- log(anc0)

bb_snp_pred <- bb_snp_pred %>% 
    mutate(t1=' [', t2=',', t3=']') %>%
    mutate(fit1 = log(fit), 
           lwr1 = log(lwr), 
           upr1 = log(upr)) %>%
    mutate(fit2 = sprintf('%.2f', fit*1e9), 
           lwr2 = sprintf('%.2f', lwr*1e9), 
           upr2 = sprintf('%.2f', upr*1e9)) %>%
    tidyr::unite('tr', fit2,t1,lwr2,t2,upr2,t3, sep = '', remove = FALSE) %>%
    tidyr::unite('CI', t1, lwr2, t2, upr2, t3, sep = '', remove = FALSE) %>%
    mutate(t1=NULL, t2=NULL, t3=NULL)

tip_pred <- bb_snp_pred[!Name %like% '^n\\d+']

bb_snp_aggre <- bb_snp[,.(nn=sum(.SD$Events), NN=sum(.SD$Callable)/1e9), by=Name]

tip_pred <- left_join(tip_pred, bb_snp_aggre, by='Name') %>%
    mutate(NN2=sprintf('%.2f', NN))


f1 <- ggtree(my_phylo_simple) + 
    geom_tiplab(align=TRUE, linetype='dashed', linesize=.6, offset = 50) +
    xlim_tree(1000) +
    theme_tree2() 

f1 <- revts(f1)

f12 <- facet_plot(f1, panel='count', data=tip_pred,
                  geom_text2,
                  mapping=aes(x=1, label=as.character(nn)),
                  hjust=1)

f12_2 <- facet_plot(f12, panel='count', data=tip_pred,
                    geom_text2,
                    mapping=aes(x=3, label=as.character(NN2)),
                    hjust=1) + xlim_expand(c(0, 3.5), 'count') 

f123 <- facet_plot(f12_2, panel='interval', data=tip_pred,
                   geom_linerange,
                   mapping = aes(x=fit1, xmin=lwr1, xmax=upr1))

f123_2 <- facet_plot(f123, panel='interval', data=tip_pred,
                     geom_vline,
                     mapping = aes(xintercept=log_anc0$fit[1]),
                     linetype = 'dashed')

f1234 <- facet_plot(f123_2, panel='interval', data=tip_pred,
                    geom_point,
                    mapping = aes(x=fit1))

f12345 <- facet_plot(f1234, panel='mean', data=tip_pred,
                     geom_text2,
                     mapping = aes(x=1, label=fit2),
                     hjust=1) + xlim_expand(c(0,0), 'mean') 

f_all <- facet_plot(f12345, panel='ci', data=tip_pred,
                    geom_text2,
                    mapping = aes(x=1, label=CI),
                    hjust=1) + xlim_expand(c(0.85, 1.02), 'mean') 

gt = ggplot_gtable(ggplot_build(f_all))
gtable_show_layout(gt) 
#gt # see plot layout in table format
f2_idx <- gt$layout$l[grep('panel-1-2', gt$layout$name)] 
gt$widths[f2_idx] = 0.55*gt$widths[f2_idx] 

f3_idx <- gt$layout$l[grep('panel-1-3', gt$layout$name)] 
gt$widths[f3_idx] = 0.8*gt$widths[f3_idx]

f4_idx <- gt$layout$l[grep('panel-1-4', gt$layout$name)] 
gt$widths[f4_idx] = 0.2*gt$widths[f4_idx] 

f5_idx <- gt$layout$l[grep('panel-1-5', gt$layout$name)] 
gt$widths[f5_idx] = 0.7*gt$widths[f5_idx]

pdf('tt.pdf', width = 10, height = 10)
grid.draw(gt)
dev.off()


### get node estimates and update the figure manually

node_sets <- data.table('clade' = c("primates_node","mammals_node","fish_node",
                                    "arthropods_node","plants_node","unicellular_node"),
                        'Name' = c('n98','n96','n53','n88','n31','n141'))

left_join(node_sets, bb_snp_pred, by='Name') %>% select(clade, fit2, lwr2, upr2)
```


We add 'Method' as a fixed effect for the above model

```{r}

####
# Add Method as a fixed term

vv <- diag(3) 
diag(vv) <- c(1e6, 1e-8,1e6)
prior_m1.2 <- list(B=list(mu=c(0,1,0), V=vv),
                  G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100),
                         G2=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))

m1.2 <- MCMCglmm(Events ~ log_callable + Method,
                 random= ~ Name + Name_copy,
                 data=bb_snp,
                 ginverse=list(Name=InverseTree_simple),
                 prior = prior_m1.2,
                 nitt=5001000, thin=1000, burnin=10000,
                 family = "poisson", verbose = FALSE,
                 pr=TRUE)

summary(m1.2)
```


We compare PO amd MA methods for Dmel and Mmus as both methods have been applied
to the two species.


For Dmel

```{r}
# Dmel

bb_dmel <- bb_snp %>% filter(Species == 'Drosophila melanogaster')

bb_dmel[is.na(u_lower), u_lower:=apply(.SD,1,function(x) inferLowerCI(x[1],x[2])), 
        .SDcols=c('Events','Callable')]

bb_dmel[is.na(u_upper), u_upper:=apply(.SD,1,function(x) inferUpperCI(x[1],x[2])), 
        .SDcols=c('Events','Callable')]

keep_rows = c('PID','PID2','u_mean', 'u_lower', 'u_upper', 'Events', 'Callable', 
              'Method')

bb_dmel <- bb_dmel %>% select(all_of(keep_rows)) %>% arrange(Method) %>%
    mutate(rid=10:1) %>% 
    mutate(Callable = sprintf('%.0f', Callable)) %>%
    mutate(u0=sprintf('%.2f',u_mean*1e9)) %>%
    mutate(u1=sprintf('%.2f',u_lower*1e9)) %>%
    mutate(u2=sprintf('%.2f',u_upper*1e9)) %>%
    mutate(t1='  [', t2=', ', t3=']') %>% 
    tidyr::unite(uCI, u0,  t1, u1, t2, u2, t3, sep = '', remove = FALSE) %>%
    mutate(t1=NULL, t2=NULL, t3=NULL)


## Plot Dmel 

f1 <- ggplot() + geom_text(aes(x=1,y=rid, label=PID2), data=bb_dmel, hjust=0) +
    theme_void()

f2 <- ggplot() + geom_text(aes(x=1,y=rid, label=Events), data=bb_dmel, hjust=1) +
    theme_void() +
    theme(plot.margin = unit(c(0,0,0,0), "cm"))

f3 <- ggplot() + geom_text(aes(x=1,y=rid, label=Callable), data=bb_dmel, hjust=1) +
    theme_void() +
    theme(plot.margin = unit(c(0,0,0,0), "cm"))

ma_mean = mean(bb_dmel[Method=='MA', u_mean])
po_mean = mean(bb_dmel[Method=='PO', u_mean])

f4 <- ggplot(data=bb_dmel) + geom_point(aes(x=u_mean, y=rid)) +
    geom_vline(xintercept = c(ma_mean, po_mean),  linetype='dashed', col='red') +
    geom_linerange(aes(xmin=u_lower, xmax=u_upper, y=rid)) +
    theme_void() + 
    theme(axis.line.x = element_line(), 
          axis.text.x = element_text(), 
          axis.title.x = element_text()) +
    scale_x_continuous(name='mutation rate(1e-9)', 
                       breaks=c(2e-9,4e-9,6e-9,8e-9), 
                       labels = c(2,4,6,8))

f5 <- ggplot() + geom_text(aes(x=1,y=rid, label=uCI), data=bb_dmel, hjust=1) +
    theme_void() 

plot_grid(f1,f2,f3, f4, f5, nrow=1, axis='bt', align='hv', 
          rel_widths = c(0.4,0.2,0.3,0.3,0.4))

ggsave('dmel_MA_PO.pdf', width = 12, height = 5)

tt <- bb_dmel[,.(ee=sum(.SD[,Events]),nn=sum(as.numeric(.SD[,Callable]))),by=Method]
binconf(tt$ee,tt$nn)*1e9

```



For Mmus

```{r}

###
# mice

bb_mus <- bb_snp %>% filter(Species == 'Mus musculus')

bb_mus[is.na(u_lower), u_lower:=apply(.SD,1,function(x) inferLowerCI(x[1],x[2])), 
       .SDcols=c('Events','Callable')]

bb_mus[is.na(u_upper), u_upper:=apply(.SD,1,function(x) inferUpperCI(x[1],x[2])), 
       .SDcols=c('Events','Callable')]

bb_mus <- bb_mus %>% select(all_of(keep_rows)) %>% arrange(Method) %>%
    mutate(rid=4:1) %>% mutate(Callable = sprintf('%.0f', Callable)) %>%
    mutate(u0=sprintf('%.2f',u_mean*1e9)) %>%
    mutate(u1=sprintf('%.2f',u_lower*1e9)) %>%
    mutate(u2=sprintf('%.2f',u_upper*1e9)) %>%
    mutate(t1='  [', t2=', ', t3=']') %>% 
    tidyr::unite(uCI, u0,  t1, u1, t2, u2, t3, sep = '', remove = FALSE) %>%
    mutate(t1=NULL, t2=NULL, t3=NULL)

## Plot mice
f1 <- ggplot() + geom_text(aes(x=1,y=rid, label=PID2), data=bb_mus, hjust=0) +
    theme_void()

f2 <- ggplot() + geom_text(aes(x=1,y=rid, label=Events), data=bb_mus, hjust=1) +
    theme_void() +
    theme(plot.margin = unit(c(0,0,0,0), "cm"))

f3 <- ggplot() + geom_text(aes(x=1,y=rid, label=Callable), data=bb_mus, hjust=1) +
    theme_void() +
    theme(plot.margin = unit(c(0,0,0,0), "cm"))

ma_mean = mean(bb_mus[Method=='MA', u_mean])
po_mean = mean(bb_mus[Method=='PO', u_mean])

f4 <- ggplot(data=bb_mus) + geom_point(aes(x=u_mean, y=rid)) +
    geom_vline(xintercept = po_mean,  linetype='dashed', col='red') +
    geom_linerange(aes(xmin=u_lower, xmax=u_upper, y=rid)) +
    theme_void() + 
    theme(axis.line.x = element_line(), 
          axis.text.x = element_text(), 
          axis.title.x = element_text()) +
    scale_x_continuous(name='mutation rate(1e-9)', 
                       breaks=c(2e-9,4e-9,6e-9,8e-9), 
                       labels = c(2,4,6,8))

f5 <- ggplot() + geom_text(aes(x=1,y=rid, label=uCI), data=bb_mus, hjust=1) +
    theme_void() 

plot_grid(f1,f2,f3, f4, f5, nrow=1, axis='bt', align='hv', 
          rel_widths = c(0.4,0.2,0.3,0.3,0.4))

ggsave('mus_MA_PO.pdf', width = 12, height = 3)


tt <- bb_mus[,.(ee=sum(.SD[,Events]),nn=sum(as.numeric(.SD[,Callable]))),by=Method]
binconf(tt$ee,tt$nn)*1e9

```


# Part Three

We are trying to explore the relationship between mutation rate and generation time,
genome size, population size and nucleotide diversity.

> Species2: species with phylogentic tree available
> Gs_Mb: genome size in MB (from NCBI)
> Gt_Year: generation time in years (from literature)
> log10_popsize: census population size in log10 scale (from Buffalo, Elife)
> diversity: synonymous or intergenic nucleotide diversity pi (from Buffalo and literautre)


```{r}

ff <- read_excel('mutation_rate_literature_updating.xlsx', sheet='gs_gt') %>%
    rename('Species2'='Species') %>%
    mutate(log_popsize = log(10^log10_popsize)) %>%
    as.data.table

cc <- left_join(bb_snp, ff, by='Species2')
cc[Species=='Homo sapiens', Group:='human']

InverseTree <- inverseA(my_phylo,scale=TRUE)$Ainv

```


## Generation Time


```{r}

cc1 <- cc[Gt_Year!=0] %>% mutate(log_Gt = log(Gt_Year)) %>%
    select(Species2, Events, log_callable, log_Gt, Group, u_mean)
cc1$Species2_phylo <- stringr::str_replace_all(cc1$Species2, ' ' ,'_')

vv <- diag(3)
diag(vv) <- c(1e6, 1e-8,1e6)
prior_m3.1 <- list(B=list(mu=c(0,1,0), V=vv),
                  G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))


m3.1 <- MCMCglmm(Events ~ log_callable + log_Gt, 
                 random= ~ Species2_phylo,
                 data=cc1, 
                 ginverse=list(Species2_phylo=InverseTree),
                 prior = prior_m3.1,
                 nitt=5001000, thin=1000, burnin=10000,
                 family = "poisson", 
                 pr=TRUE, verbose = FALSE)



# prior_m3.1.x <- list(B=list(mu=c(0,1,0), V=vv),
#                   G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100),
#                          G2=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))
# 
# m3.1.x <- MCMCglmm(Events ~ log_callable + log_Gt, 
#                  random= ~ Species2_phylo + Species2,
#                  data=cc1, 
#                  ginverse=list(Species2_phylo=InverseTree),
#                  prior = prior_m3.1.x,
#                  nitt=1001000, thin=1000, burnin=10000,
#                  family = "poisson",
#                  pr=TRUE, verbose = FALSE)


## Create prediction table conditional on all random levels
cc1_pred <- data.table()
for(ss in unique(cc1$Species2_phylo)){
    # print(ss)
    gg <- unique(cc1[Species2_phylo==ss, Group])
    tmp <- data.table('log_Gt' = seq(-5.29,3.36,0.01),
                      'Events' = 0,
                      'log_callable' = 0,
                      'Species2_phylo' = ss)
    tmp[1,log_callable:=1]
    tmp <- predict(m3.1, newdata = tmp,  type='response', marginal = NULL) %>%
        cbind(tmp,.) %>% filter(log_callable==0) %>%
        select(log_Gt,Species2_phylo,V1) %>%
        mutate(Group=gg)
    cc1_pred <- rbind(cc1_pred, tmp)

}



scol_1 <- c('human'=human_col, 'primates'=primates_col, 'mammals'='gold',
            'birds'=birds_col, 'fish'=fish_col, 'arthropods'=arthropods_col,
            'worm'=worm_col,'plants'=plants_col, 'fungus'=fungus_col)


## log-scale y-axis
y_breaks = c(-26,-24,-22,-20,-18,-16,-14)

P_gt <- ggplot(data=cc1,aes(x=log_Gt, y=log(u_mean))) +
    geom_point(aes(col=Group), size=3) +
    scale_x_continuous(name='Generation time (Days)',
                       breaks=c(-4,-2,0,2),
                       labels = c('7','50','365','2700')) +
    scale_y_continuous(name='SNM rates', limits = range(y_breaks),
                       breaks = y_breaks,
                       labels=sprintf('%.1e', exp(y_breaks))) +
    scale_color_manual(values=scol_1) +
    geom_smooth(method='lm', se=FALSE, size=1, col='black') +
    geom_line(aes(x=log_Gt, y=log(V1), col=Group, group=Species2_phylo),
              data=cc1_pred, alpha=0.3) +
    theme_classic() +
    theme(legend.position = 'none',
          axis.text = element_text(size=13),
          axis.title = element_text(size=15))


```





## Genome Size


```{r}

cc2 <- cc[Gs_Mb!=0] %>% mutate(log_Gs_Mb = log(Gs_Mb))  %>%
    select(Species2, Events, log_callable, log_Gs_Mb, Group, u_mean)

cc2$Species2_phylo <- stringr::str_replace_all(cc2$Species2, ' ' ,'_')

vv <- diag(3)
diag(vv) <- c(1e6, 1e-8,1e6)
prior_m3.2 <- list(B=list(mu=c(0,1,0), V=vv),
                  G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))

m3.2 <- MCMCglmm(Events ~ log_callable + log_Gs_Mb, 
                 random= ~ Species2_phylo,
                 data=cc2, 
                 ginverse=list(Species2_phylo=InverseTree),
                 prior = prior_m3.2,
                 nitt=5001000, thin=1000, burnin=10000,
                 family = "poisson",
                 pr=TRUE, verbose = FALSE)


# prior_m3.2.x <- list(B=list(mu=c(0,1,0), V=vv),
#                   G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100),
#                          G2=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))
# 
# m3.2.x <- MCMCglmm(Events ~ log_callable + log_Gs_Mb, 
#                  random= ~ Species2_phylo + Species2,
#                  data=cc2, 
#                  ginverse=list(Species2_phylo=InverseTree),
#                  prior = prior_m3.2.x,
#                  nitt=5001000, thin=1000, burnin=10000,
#                  family = "poisson",
#                  pr=TRUE, verbose = FALSE)






cc2_pred <- data.table()

for(ss in unique(cc2$Species2_phylo)){
    #print(ss)
    gg <- unique(cc2[Species2_phylo==ss, Group])
    tmp <- data.table('log_Gs_Mb' = seq(2.19,8.04, 0.01),
                      'Events' = 0,
                      'log_callable' = 0,
                      'Species2_phylo' = ss)
    tmp[1,log_callable:=1]
    tmp <- predict(m3.2, newdata = tmp,  type='response', marginal = NULL) %>%
        cbind(tmp,.) %>% filter(log_callable==0) %>%
        select(log_Gs_Mb,Species2_phylo,V1) %>%
        mutate(Group=gg)
    cc2_pred <- rbind(cc2_pred, tmp)

}


scol_2 <- c('human'=human_col, 'primates'=primates_col, 'mammals'='gold',
            'birds'=birds_col, 'fish'=fish_col, 'arthropods'=arthropods_col,
            'worm'=worm_col, 'plants'=plants_col, 'fungus'=fungus_col,
            'unicellular'=unicellular_col)

x_breaks3 = c(2,4,6,8)

P_gs <- ggplot(data=cc2, aes(x=log_Gs_Mb, y=log(u_mean))) +
    geom_point(aes(col=Group), size=3) +
    scale_x_continuous(name='Genome size (Mb)',
                       breaks = x_breaks3,
                       labels = sprintf('%.1f', exp(x_breaks3))) +
    scale_y_continuous(name='SNM rates',
                       limits = range(y_breaks),
                       breaks = y_breaks,
                       labels = sprintf('%.1e', exp(y_breaks))) +
    scale_color_manual(name='', values=scol_2) +
    geom_smooth(method='lm', se=FALSE, size=1, col='black') +
    geom_line(aes(x=log_Gs_Mb, y=log(V1), col=Group, group=Species2_phylo),
              data=cc2_pred, alpha=0.3) +
    theme_classic() +
    theme(legend.position = 'none',
          axis.text = element_text(size=13),
          axis.title = element_text(size=15))


```









## Population Size

```{r}

cc3 <- cc[log_popsize!=0]  %>%
    select(Species2, Events, log_callable, log_popsize, Group, u_mean)

cc3$Species2_phylo <- stringr::str_replace_all(cc3$Species2, ' ' ,'_')

vv <- diag(3)
diag(vv) <- c(1e6, 1e-8,1e6)
prior_m3.3 <- list(B=list(mu=c(0,1,0), V=vv),
                  G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))


m3.3 <- MCMCglmm(Events ~ log_callable + log_popsize, 
                 random= ~ Species2_phylo,
                 data=cc3, 
                 ginverse=list(Species2_phylo=InverseTree),
                 prior = prior_m3.3,
                 nitt=5001000, thin=1000, burnin=10000,
                 family = "poisson",
                 pr=TRUE, verbose = FALSE)


cc3_pred <- data.table()

for(ss in unique(cc3$Species2_phylo)){
    #print(ss)
    gg <- unique(cc3[Species2_phylo==ss, Group])
    tmp <- data.table('log_popsize' = seq(12.66,34.95,0.01),
                      'Events' = 0,
                      'log_callable' = 0,
                      'Species2_phylo' = ss)
    tmp[1,log_callable:=1]
    tmp <- predict(m3.3, newdata = tmp,  type='response', marginal = NULL) %>%
        cbind(tmp,.) %>% filter(log_callable==0) %>%
        select(log_popsize,Species2_phylo,V1) %>%
        mutate(Group=gg)
    cc3_pred <- rbind(cc3_pred, tmp)
}



scol_3 <- c('human'=human_col, 'primates'=primates_col, 'mammals'='gold',
            'birds'=birds_col, 'arthropods'=arthropods_col, 'worm'=worm_col)

x_breaks2 = c(15,20,25,30,35)

P_ps <- ggplot(data=cc3,aes(x=log_popsize, y=log(u_mean))) +
    geom_point(aes(col=Group), size=3) +
    scale_x_continuous(name='Population size' ,
                       breaks = x_breaks2,
                       labels=sprintf('%.1e', exp(x_breaks2))) +
    scale_y_continuous(name='SNM rates',
                       limits = range(y_breaks),
                       breaks = y_breaks,
                       labels=sprintf('%.1e', exp(y_breaks))) +
    scale_color_manual(name='', values=scol_2) +
    geom_smooth(method='lm', se=FALSE, size=1, col='black') +
    geom_line(aes(x=log_popsize, y=log(V1), col=Group, group=Species2_phylo),
              data=cc3_pred, alpha=0.3) +
    theme_classic() +
    theme(legend.position = c(0.53,0.15),
          legend.text = element_text(size=12),
          axis.text = element_text(size=13),
          axis.title = element_text(size=15)) +
    guides(color=guide_legend(nrow=3, byrow=FALSE))



```


If we fit all three factors together: generation time, genome size and population size


```{r}
# all factors together

vv <- diag(5)
diag(vv) <- c(1e6, 1e-8,1e6,1e6,1e6)
prior_m.ccx <- list(B=list(mu=c(0,1,0,0,0), V=vv),
                  G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))

ccx <- cc[Gt_Year!=0 & log_popsize !=0] %>%
    mutate(log_Gt_Year = log(Gt_Year), log_Gs_Mb = log(Gs_Mb)) %>%
    select(Species2, Events, log_callable, log_popsize, log_Gt_Year, log_Gs_Mb,
            Group, u_mean)

ccx$Species2_phylo <- stringr::str_replace_all(cc3$Species2, ' ' ,'_')



m.ccx <- MCMCglmm(Events ~ log_callable + log_Gs_Mb + log_popsize + log_Gt_Year, 
                 random= ~ Species2_phylo,
                 data=ccx, 
                 ginverse=list(Species2_phylo=InverseTree),
                 prior = prior_m.ccx,
                 family = 'poisson',
                 nitt=5001000, thin=1000, burnin=10000,
                 pr=TRUE, verbose = FALSE)



```



## Nucleotide Diversity


Use the model as Jarrod suggested to explore the relationship between nucleotide
diversity and mutation rate. 


```{r}

# get predicted mutation rate based on mu ~ (1|phylo)

bb_snp$Species2_phylo = stringr::str_replace_all(bb_snp$Species2, ' ', '_') 

vv <- diag(2) 
diag(vv) <- c(1e6, 1e-8)
prior_m4.1 <- list(B=list(mu=c(0,1), V=vv),
                 G=list(G2=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))


m4.1 <- MCMCglmm(Events ~ log_callable,
               random= ~ Species2_phylo,
               data=bb_snp,
               ginverse=list(Species2_phylo=InverseTree),
               prior = prior_m4.1,
               nitt=5001000, thin=1000, burnin=10000,
               family = "poisson",
               pr=TRUE, verbose = FALSE)

m4.1_pred <- data.table('Species2_phylo'=c(my_phylo$node.label, 
                                             my_phylo$tip.label)) %>%
    filter(Species2_phylo!='n') %>%
    mutate(Events=0) %>%
    mutate(log_callable=0)

m4.1_pred[Species2_phylo=='n9', log_callable:=1]
m4.1_pred$Species2 <- m4.1_pred$Species2_phylo
m4.1_pred[Species2_phylo %like% '^n\\d+', Species2:=NA]
m4.1_pred$Species2 <- stringr::str_replace_all(m4.1_pred$Species2,'_', ' ')



m4.1_pred <- predict(m4.1, interval="confidence", newdata=m4.1_pred, 
                       type='response', marginal = NULL) %>%
    cbind(m4.1_pred,.)

mu_pred <- m4.1_pred[!is.na(Species2), c('fit','Species2_phylo', 'Species2')]
div <- cc[diversity!=0, c('Species2', 'diversity')]

mu_pred_div <- left_join(div, mu_pred, by = 'Species2') %>%
    mutate(log_div = log(diversity)) %>%
    mutate(log_mu = log(fit)) %>%
    mutate(diversity=NULL, fit=NULL) %>% unique()



##
# create data table for model
d1 <- bb_snp %>% select(Species2_phylo, Events, log_callable) %>%
    mutate(trait = 'mu') %>%
    mutate(family = 'poisson') %>%
    rename(y = Events)

d2 <- mu_pred_div %>% select(Species2_phylo, log_div) %>%
    rename(y = log_div) %>%
    mutate(trait = 'div') %>%
    mutate(family = 'gaussian') %>%
    mutate(log_callable = 0) %>%
    select(Species2_phylo, y, log_callable, trait, family)

d12 <- rbind(d1,d2)
d12$trait <- factor(d12$trait, levels = c('mu', 'div'))


## fit bivariate model
prior_m4.2 <- list(R=list(V=diag(2), nu=0.002),
            G=list(G1=list(V=diag(2), nu=2, alpha.V=diag(2)*100, beta.mu=0, beta.V=10)))

# m4.2_biv <- MCMCglmm(y ~ trait + at.level(trait, 'mu'):log_callable, 
#                      random = ~ante1(trait):Species2_phylo, 
#                      rcov = ~idh(trait):units,
#                      data = d12, 
#                      ginverse = list(Species2_phylo=InverseTree),
#                      prior = prior_m4.2,
#                      nitt=1001000, thin=1000, burnin=10000,
#                      family = NULL, verbose = FALSE
#                      )

m4.2_biv.x <- MCMCglmm(y ~ trait - 1 + at.level(trait, 'mu'):log_callable,
                     random = ~ante1(trait):Species2_phylo,
                     rcov = ~idh(trait):units,
                     data = d12,
                     ginverse = list(Species2_phylo=InverseTree),
                     prior = prior_m4.2,
                     nitt=5001000, thin=1000, burnin=10000,
                     family = NULL, verbose = FALSE
                     )

plot(posterior.ante(m4.2_biv.x$VCV[,1:4], k=1))

posterior.ante(m4.2_biv.x$VCV[,1:4], k=1)[,3] %>% HPDinterval()

posterior.ante(m4.2_biv.x$VCV[,1:4], k=1)[,3] %>% mean
# p-value
sum(posterior.ante(m4.2_biv.x$VCV[,1:4], k=1)[,3] > 1) / nrow(m4.2_biv.x$VCV)






# add a regression line
i_mu <- summary(m4.2_biv.x)$solutions[1]
i_div <- summary(m4.2_biv.x)$solutions[2]
b_slope <- posterior.ante(m4.2_biv.x$VCV[,1:4], k=1) %>% colMeans() %>% .[[3]]
b_slope2 = 1
plot_df <- data.table(plot_mu = seq(min(mu_pred_div$log_mu)-1,
                                    max(mu_pred_div$log_mu)+1,
                                    0.01),
                      plot_div = 0,
                      plot_div2 = 0)

for(i in 1:nrow(plot_df)){
    imu <- plot_df[i,plot_mu]
    y_pred <- i_div + b_slope*(imu - i_mu)
    y_pred2 <- i_div + b_slope2*(imu - i_mu)
    plot_df[i, plot_div:=y_pred]
    plot_df[i, plot_div2:=y_pred2]
}




# add group for colour aesthetic

mu_pred_div <- left_join(mu_pred_div, unique(bb_snp[,c('Species2','Group')]), by='Species2')


scol_4 <- c('human'=human_col, 'primates'=primates_col, 'mammals'='gold',
            'birds'=birds_col, 'fish'=fish_col, 'arthropods'=arthropods_col,
            'worm'=worm_col, 'plants'=plants_col, 'fungus'=fungus_col,
            'unicellular'=unicellular_col)


## put mutation rate on Y-axis, pi on X-axis
mu_pred_div[Species2_phylo=='Homo_sapiens', Group:='human']

x_breaks3 = c(-25,-23,-21,-19,-17)
y_breaks3 = seq(-10,0,1)


P_div <- ggplot(data=mu_pred_div,aes(x=log_mu, y=log_div)) +
    geom_point(aes(col=Group), size=3) +
    scale_y_continuous(name='Nucleotide diversity (%)',
                       limits = c(-9.2,-0.6),
                       breaks = y_breaks3,
                       labels=sprintf('%.2f', exp(y_breaks3)*100)) +
    scale_x_continuous(name='Predicted SNM rates',
                       limits = range(x_breaks3),
                       breaks = x_breaks3,
                       labels=sprintf('%.1e', exp(x_breaks3))) +
    geom_line(aes(x=plot_mu, y=plot_div),
              data=plot_df, size=2, colour='grey', alpha=0.5) +
    geom_line(aes(x=plot_mu, y=plot_div2),
              data=plot_df, size=2, colour='grey', alpha=0.5, linetype='dashed') +
    scale_color_manual(name='', values=scol_4) +
    theme_classic() +
    theme(legend.position = 'none',# c(0.85,0.4),
          axis.text = element_text(size=13),
          axis.title = element_text(size=15),
          legend.background = element_rect(fill="white",
                                  size=0.5, linetype="solid", 
                                  colour ="black"))




## add generation time as a fixed effect

d12.y <- cc1[,c('Species2_phylo','log_Gt')] %>% unique() %>% 
    left_join(d12, ., by='Species2_phylo') %>%
    filter(!is.na(log_Gt))


Extended_biv_model <- MCMCglmm(y ~ trait - 1 + trait:log_Gt + 
                                   at.level(trait, 'mu'):log_callable ,
                     random = ~ante1(trait):Species2_phylo, 
                     rcov = ~idh(trait):units,
                     data = d12.y, 
                     ginverse = list(Species2_phylo=InverseTree),
                     prior = prior_m4.2,
                     nitt=5001000, thin=1000, burnin=10000,
                     family = NULL, verbose = FALSE
                     )

summary(Extended_biv_model)
plot(posterior.ante(Extended_biv_model$VCV[,1:4], k=1))

posterior.ante(Extended_biv_model$VCV[,1:4], k=1)[,3] %>% HPDinterval()

posterior.ante(Extended_biv_model$VCV[,1:4], k=1)[,3] %>% mean


sum(posterior.ante(Extended_biv_model$VCV[,1:4], k=1)[,3] > 1) / nrow(Extended_biv_model$VCV)


```



Put all plots together

```{r}
plot_grid(P_gt,P_gs, P_ps, P_div, ncol=2, 
          labels = c('(A)', '(B)', '(C)', '(D)'), label_x = 0.2,
          align = 'hv')

ggsave('mu_genomeSize_generationTime_new.pdf', width = 11, height=10, units = 'in')

```





# Part Four

## INDELs
INDELs: we extract indels from the dataset and do some filterings as we did to the 
SNMs.


```{r}
dd <- bb %>% extract(CITATION, 'Author', '^(.*?),') %>%
    tidyr::unite('PID2','Author','Date', sep=',') %>%
    filter(TYPE=='indel') %>% 
    filter(Unit=='bp/generation') %>%
    filter(!is.na(u_mean)) %>%
    mutate(u_mean = as.numeric(u_mean)) %>% 
    mutate(u_lower = as.numeric(u_lower)) %>%
    mutate(u_upper = as.numeric(u_upper)) %>%
    mutate(SE = as.numeric(SE)) %>%
    mutate(Events = as.integer(Events)) %>%
    filter(!Comments %like% 'exome') %>%
    filter(!Comments %like% 'male') %>%
    filter(!Comments %like% 'sperm') %>%
    select(c('PID', 'PID2','u_mean','u_lower','u_upper','Events','Callable', 'SE',
             'Name','Species','Method', 'Group')) %>%
    as.data.table

# Calculate interval if SE provided
dd[is.na(Events) & is.na(u_lower) & !is.na(SE) & !is.na(u_mean), 
       u_lower:=u_mean-1.96*SE]
dd[is.na(Events) & is.na(u_upper) & !is.na(SE) & !is.na(u_mean), 
       u_upper:=u_mean+1.96*SE]

# estimate Events based on interval
dd[is.na(Events) & !is.na(u_lower), 
   Events:=apply(.SD,1,function(x) inferEvents(x[1],x[2],x[3])), 
   .SDcols=c('u_mean','u_lower','u_upper')]
# Assume 1 Event if interval no available
dd[is.na(Events) & is.na(u_lower), Events:=1]
dd[, Callable:=Events/u_mean]
dd$PID <- as.character(dd$PID)

# dd$Species %>% unique() %>% length

dd[,Species2:=Species]
dd[Species=='Amphilophus', Species2:='Amphilophus labiatus']
dd[Species=='Anopheles stephensi', Species2:='Anopheles gambiae']
dd[Species=='Anopheles coluzzii', Species2:='Anopheles gambiae']
dd[Species %in% c('Paramecium sexaurelia','Paramecium biaurelia'), Species2:='Paramecium tetraurelia']
dd[Species=='Chlamydomonas incerta', Species2:='Chlamydomonas reinhardtii']
dd[Species=='Daphnia galeata', Species2:='Daphnia dubia']
dd[Species %in% c('Ostreococcus mediterraneus', 'Bathycoccus prasinos'), Species2:='Ostreococcus tauri']
dd[Species=='Saccharomycodes ludwigii', Species2:='Hanseniaspora valbyensis']
# following automatically replaced with neighbouring sppecies by treetime.org
dd[Species=='Micromonas pusilla', Species2:='Micromonas']
dd[Species=='Picochlorum costavermella', Species2:='Prototheca wickerhamii']
dd[Species=='Rhodotorula toruloides', Species2:='Rhodotorula graminis']
dd[Species=='Bombus terrestris', Species2:='Bombus']
dd[Species=='Chironomus riparius', Species2:='Chironomus pallidivittatus']
dd[Species=='Pristionchus pacificus', Species2:='Pristionchus']

dd_unique_species <- dd[,"Species2"] %>% unique()

```

We extract and prune tree for indels.


```{r}
# create phylogeny based on species common names

dd[Group=='human', Group:='primates']
dd_name <- dd[,.SD[1,], by='Name', .SDcols=c('Species2')] 
dd_name$Species2_phylo <- stringr::str_replace_all(dd_name$Species2, ' ', '_')

my_phylo_simple.dd <- ape::keep.tip(my_phylo, tip=dd_name$Species2_phylo)
my_phylo_simple.dd <- ladderize(my_phylo_simple.dd, right = FALSE)


for(nn in my_phylo_simple.dd$tip.label){
    nn2 <- dd_name[Species2_phylo==nn, Name]
    my_phylo_simple.dd$tip.label <- 
        stringr::str_replace_all(my_phylo_simple.dd$tip.label, nn, nn2)
}

InverseTree_simple.dd <- inverseA(my_phylo_simple.dd,scale=TRUE)$Ainv
```


Fit model for indels: mu ~ (1|phylogeny)

```{r}
dd$log_callable <- log(dd$Callable)
vv <- diag(2) 
diag(vv) <- c(1e6, 1e-8)
prior_m5.1 <- list(B=list(mu=c(0,1), V=vv),
                 G=list(G1=list(V=1, nu=1, alpha.mu=0, alpha.V=100)))

m5.1 <- MCMCglmm(Events ~ log_callable, 
               random= ~ Name,
               data=dd, 
               ginverse=list(Name=InverseTree_simple.dd),
               prior = prior_m5.1,
               nitt=5001000, thin=1000, burnin=10000,
               family = "poisson",
               pr=TRUE, verbose = FALSE)

dd_pred <- data.table('Name'=c(my_phylo_simple.dd$node.label, 
                               my_phylo_simple.dd$tip.label)) %>% 
    filter(Name!='n') %>% 
    mutate(Events=0) %>%
    mutate(log_callable=ifelse(Name=='n86', 1, 0)) 

dd_pred <- predict(m5.1, marginal=NULL, interval="confidence",newdata=dd_pred) %>%
    cbind(dd_pred,.)

```


Create data table for plot

```{r}

### get ancestor node 
anc0.dd <- dd_pred[Name=='n130',4:6]
log_anc0.dd <- log(anc0.dd)

dd_pred <- dd_pred %>% 
    mutate(t1=' [', t2=',', t3=']') %>%
    mutate(fit1 = log(fit), 
           lwr1 = log(lwr), 
           upr1 = log(upr)) %>%
    mutate(fit2 = sprintf('%.2f', fit*1e9), 
           lwr2 = sprintf('%.2f', lwr*1e9), 
           upr2 = sprintf('%.2f', upr*1e9)) %>%
    tidyr::unite('tr', fit2, t1, lwr2, t2, upr2,t3, sep = '', remove = FALSE) %>%
    tidyr::unite('CI',  t1, lwr2, t2, upr2,t3, sep = '', remove = FALSE) %>%
    mutate(t1=NULL, t2=NULL, t3=NULL)

tip_pred.dd <- dd_pred[!Name %like% '^n[0-9]']

dd_aggre <- dd[,.(nn=sum(.SD$Events), NN=sum(.SD$Callable)/1e9), by=Name]

tip_pred.dd <- left_join(tip_pred.dd, dd_aggre, by='Name') %>%
    mutate(NN2=sprintf('%.2f', NN))

```


Plot phylogentic forest tree for indels.


```{r}
## Plot
f1 <- ggtree(my_phylo_simple.dd) + 
    geom_tiplab(align=TRUE, linetype='dashed', linesize=.6, offset = 0) +
    xlim_tree(1000) +
    theme_tree2() 
f1 <- revts(f1)

f12 <- facet_plot(f1, panel='count', data=tip_pred.dd,
                  geom_text2,
                  mapping=aes(x=1, label=as.character(nn)),
                  hjust=1)

f12_2 <- facet_plot(f12, panel='count', data=tip_pred.dd,
                    geom_text2,
                    mapping=aes(x=3, label=as.character(NN2)),
                    hjust=1) + xlim_expand(c(0, 3.5), 'count') 

f123 <- facet_plot(f12_2, panel='interval', data=tip_pred.dd,
                   geom_linerange,
                   mapping = aes(x=fit1, xmin=lwr1, xmax=upr1))

f123_2 <- facet_plot(f123, panel='interval', data=tip_pred.dd,
                     geom_vline,
                     mapping = aes(xintercept=log_anc0.dd$fit[1]),
                     linetype = 'dashed')

f1234 <- facet_plot(f123_2, panel='interval', data=tip_pred.dd,
                    geom_point,
                    mapping = aes(x=fit1))

f12345 <- facet_plot(f1234, panel='mean', data=tip_pred.dd,
                     geom_text2,
                     mapping = aes(x=1, label=fit2),
                     hjust=1) + xlim_expand(c(0,0), 'mean') 

f_all <- facet_plot(f12345, panel='ci', data=tip_pred.dd,
                    geom_text2,
                    mapping = aes(x=1, label=CI),
                    hjust=1) + xlim_expand(c(0.85, 1.02), 'mean') 


gt = ggplot_gtable(ggplot_build(f_all))
gtable_show_layout(gt) 
#gt # see plot layout in table format
f2_idx <- gt$layout$l[grep('panel-1-2', gt$layout$name)] 
gt$widths[f2_idx] = 0.55*gt$widths[f2_idx] 

f3_idx <- gt$layout$l[grep('panel-1-3', gt$layout$name)] 
gt$widths[f3_idx] = 0.8*gt$widths[f3_idx]

f4_idx <- gt$layout$l[grep('panel-1-4', gt$layout$name)] 
gt$widths[f4_idx] = 0.2*gt$widths[f4_idx] 

f5_idx <- gt$layout$l[grep('panel-1-5', gt$layout$name)] 
gt$widths[f5_idx] = 0.7*gt$widths[f5_idx]

pdf('tt.pdf', width = 9, height = 6)
grid.draw(gt) 
dev.off()

node_sets <- data.table('clade' = c("mammals_node",
                                    "arthropods_node","plants_node"),
                        'Name' = c('n97','n88','n31'))
left_join(node_sets, dd_pred, by='Name')

```


## Correlation Indels vs SNMs


Use mcmcglmm to explore the correlation between INDELs and SNMs while accounting
for phylogeny


```{r}

dx <- bb %>% filter(TYPE %in% c('snp','indel')) %>% 
    extract(CITATION, 'Author', '^(.*?),') %>%
    tidyr::unite('PID2','Author','Date', sep=',') %>%
    filter(Unit=='bp/generation') %>%
    filter(!is.na(u_mean)) %>%
    mutate(u_mean = as.numeric(u_mean)) %>% 
    mutate(u_lower = as.numeric(u_lower)) %>%
    mutate(u_upper = as.numeric(u_upper)) %>%
    mutate(Events = as.integer(Events)) %>%
    filter(!Comments %like% 'exome') %>%
    filter(!Comments %like% 'male') %>%
    filter(!Comments %like% 'sperm') %>%
    select(c('TYPE', 'PID2','SID','u_mean','Name','Species', 'Group')) %>%
    as.data.table

dx.indel <- dx[TYPE=='indel'] %>% rename(u_indel = u_mean)
dx.snp <- dx[TYPE=='snp'] %>% rename(u_snp = u_mean)
dx_snp_indel <-  left_join(dx.indel, dx.snp[,c('SID', 'u_snp')], by='SID') %>% 
    left_join(., unique(bb_snp[,c('Species','Species2')]), by='Species') %>%
    filter(!is.na(u_snp)) %>%
    mutate(log_uindel = log(u_indel)) %>%
    mutate(log_usnp = log(u_snp)) %>%
    mutate(TYPE=NULL) %>%
    mutate(Species2_phylo = stringr::str_replace_all(Species2, ' ', '_'))


# Prior as advised by Jarrod
prior_m5.2 <- list(
    R = list(V = diag(2), nu = 1.002), 
    G = list(
        G1= list(V = diag(2) , nu = 2, alpha.mu=c(0,0),alpha.V=diag(2)*25^2)
        
    )
)

m5.2 <- MCMCglmm(cbind(log_uindel,log_usnp) ~ trait-1, 
               random= ~us(trait):Species2_phylo,
               rcov = ~idh(trait):units,
               data=dx_snp_indel,
               ginverse=list(Species2_phylo=InverseTree),
               prior = prior_m5.2,
               family = rep('gaussian',2), 
               nitt=1001000, thin=1000, burnin=10000,
               pr=TRUE, verbose = FALSE)

#The covariance is (significantly) positive
HPDinterval(m5.2$VCV[,2])
sum(m5.2$VCV[,2]<0)/nrow(m5.2$VCV)

HPDinterval(m5.2$VCV[,2]/(sqrt(m5.2$VCV[,1])*sqrt(m5.2$VCV[,4])))
mean(m5.2$VCV[,2]/(sqrt(m5.2$VCV[,1])*sqrt(m5.2$VCV[,4])))


```



Plot correlation between indels and SNMs

```{r}
XYmin = -27
XYmax = -17
xy_breaks4 = seq(XYmin,XYmax,2)


dx_snp_indel$Group <- factor(dx_snp_indel$Group,
                       levels=c('human', 'mammals', 'worm',
                                 'arthropods', 'plants', 'unicellular'))
col4 = scol_2[levels(dx_snp_indel$Group)]


ggplot() + geom_abline(intercept = 0, slope = 1, linetype='dashed', 
                       color='grey', size=1) +
    geom_point(data=dx_snp_indel, aes(x=log(u_snp), y=log(u_indel), color=Group), 
               size=5) +
    scale_x_continuous(name='SNM rates', limits = c(XYmin, XYmax), 
                       breaks = xy_breaks4, label=sprintf('%.1e',exp(xy_breaks4))) +
    scale_y_continuous(name='indel mutation rates', limits = c(XYmin, XYmax), 
                       breaks = xy_breaks4, label=sprintf('%.1e',exp(xy_breaks4))) + 
    scale_color_manual(name='', values=col4) + 
    theme_classic() +
    theme(legend.position = c(0.2,0.8), legend.text = element_text(size=18),
          axis.title = element_text(size=18),
          axis.text = element_text(size=17)
    )

ggsave('snp_indel_dots.pdf', width = 7, height = 6, units = 'in')
```


## Deletion and insertion

Generate data table for deletion and insertion.

```{r}
## deletion and insertion

di <- bb %>% extract(CITATION, 'Author', '^(.*?),') %>%
    tidyr::unite('PID2','Author','Date', sep=',') %>%
    filter(TYPE %in% c('insertion', 'deletion')) %>% 
    filter(Unit=='bp/generation') %>% 
    filter(!is.na(u_mean)) %>%
    mutate(u_mean = as.numeric(u_mean)) %>% 
    mutate(u_lower = as.numeric(u_lower)) %>%
    mutate(u_upper = as.numeric(u_upper)) %>%
    mutate(Events = as.integer(Events)) %>%
    filter(!Comments %like% 'exome') %>%
    filter(!Comments %like% 'male') %>%
    filter(!Comments %like% 'sperm') %>%
    select(c('PID', 'PID2','TYPE','SID','u_mean','u_lower','u_upper','Events',
             'Callable','Name','Species','Method', 'Group')) %>%
    as.data.table


## Assume 1 indel event where is 0
di <- di %>% mutate(Events=ifelse(Events==0,1,Events)) %>%
    mutate(Callable = Events/u_mean)


di.1 <- di %>% select('PID2', 'SID', 'Name','Species','Group', 'TYPE', 'Events') %>% 
    tidyr::spread(TYPE, Events) 
di.2 <- di %>% select('PID2', 'SID', 'Name','Species','Group', 'TYPE', 'Callable') %>% 
    tidyr::spread(TYPE, Callable) %>%
    mutate(deletion_callable = ifelse(is.infinite(deletion), insertion, deletion)) %>%
    mutate(insertion_callable = ifelse(is.infinite(insertion), deletion, insertion)) %>%
    select('SID','deletion_callable','insertion_callable')

di.12 <- left_join(di.1, di.2, by='SID') 

di.12 <- apply(di.12[,c('deletion','insertion','deletion_callable','insertion_callable')],
             1,
             function(x) DescTools::BinomRatioCI(x[1],x[3],x[2],x[4])) %>% 
    t() %>% cbind(di.12, .) %>%
    rename(c('DIrr'='V1', 'DIlr'='V2', 'DIur'='V3'))
# log transformed ratio
di.12[,DIrr:=log(DIrr)]
di.12[,DIlr:=log(DIlr)]
di.12[,DIur:=log(DIur)]

di.12[,DIrr2:=sprintf('%.2f',DIrr)]
di.12[,DIlr2:=sprintf('%.2f',DIlr)]
di.12[,DIur2:=sprintf('%.2f',DIur)]



zz <- c("midge","fruit fly","bumblebee","honeybee",
        "water flea","nematode",
        "human","mice", "cattle",
        "yeast", "peach","arabidopsis","rice",
        "algae", "amoeba")

di.123 <- di.12 %>% mutate(t1=' [', t2=',', t3=']') %>% 
    tidyr::unite('CI',  t1, DIlr2, t2, DIur2, t3, sep = '', remove = FALSE) %>%
    mutate(t1=NULL, t2=NULL, t3=NULL, Group=NULL) %>%
    mutate(Name=factor(Name, levels=zz)) %>%
    arrange(Name) %>%
    mutate(rid=1:nrow(.))
```

Plot deletion and insertion ratio by `studies` 


```{r}


f0 <- ggplot() + geom_text(aes(x=1,y=rid, label=PID2), data=di.123, hjust=1) +
    theme_void()

f1 <- ggplot() + geom_text(aes(x=1,y=rid, label=Name), data=di.123, hjust=1) +
    theme_void()

f2 <- ggplot() + geom_text(aes(x=1,y=rid, label=DIrr2), data=di.123, hjust=1) +
    theme_void() +
    theme(plot.margin = unit(c(0,0,0,0), "cm"))

f3 <- ggplot() + geom_text(aes(x=1,y=rid, label=CI), data=di.123, hjust=1) +
    theme_void() +
    theme(plot.margin = unit(c(0,0,0,0), "cm"))

f4 <- ggplot(data=di.123) + geom_point(aes(x=DIrr, y=rid)) +
    geom_vline(xintercept = 0,  linetype='dashed', col='red') +
    geom_linerange(aes(xmin=DIlr, xmax=DIur, y=rid)) +
    theme_void() + 
    theme(axis.line.x = element_line(), 
          axis.text.x = element_text(), 
          axis.title.x = element_text()) +
    scale_x_continuous(name='log ratio of deletion/insertion')

plot_grid(f0,f1,f2,f3, f4, nrow=1, axis='bt', align='hv', 
          rel_widths = c(0.4,0.2,0.1,0.2,0.3))

ggsave('deletion_insertion_ratio_full.pdf', width = 12, height = 12)
```



Plot deletion and insertion ratio by `species`.

```{r}
di.12_aggre <- di.12[,.(deletion=sum(.SD[,deletion]),
                    insertion=sum(.SD[,insertion]),
                    deletion_callable=sum(.SD[,deletion_callable]),
                    insertion_callable=sum(.SD[,insertion_callable])),
                 by=Name,
                 .SDcols=c('Name', 'deletion', 'insertion', 'deletion_callable', 
                           'insertion_callable')] %>%
    mutate(Group='log ratio and 95%CI')

di.12_aggre <- apply(di.12_aggre[,c('deletion','insertion','deletion_callable','insertion_callable')],
                   1,
                   function(x) DescTools::BinomRatioCI(x[1],x[3],x[2],x[4])) %>% 
    t() %>% as.data.table() %>%
    rename(c('DIrr'='V1', 'DIlr'='V2', 'DIur'='V3')) %>%  cbind(di.12_aggre, .) 

di.12_aggre[,DIrr1:=log(DIrr)]
di.12_aggre[,DIlr1:=log(DIlr)]
di.12_aggre[,DIur1:=log(DIur)]
di.12_aggre[,DIrr2:=sprintf('%.2f',log(DIrr))]
di.12_aggre[,DIlr2:=sprintf('%.2f',log(DIlr))]
di.12_aggre[,DIur2:=sprintf('%.2f',log(DIur))]


di.12_aggre <- di.12_aggre %>% mutate(t1=' [', t2=',', t3=']') %>% 
    tidyr::unite('CI',  t1, DIlr2, t2, DIur2, t3, sep = '', remove = FALSE) %>%
    mutate(t1=NULL, t2=NULL, t3=NULL, Group=NULL)

di.12_aggre$Name <- factor(di.12_aggre$Name, 
                         levels = rev(c("midge","fruit fly","bumblebee","honeybee",
                                         "water flea","nematode",
                                          "human","mice", "cattle",
                                          "yeast", "peach","arabidopsis","rice",
                                          "algae", "amoeba")))


f1 <- ggplot() + geom_text(aes(x=1,y=Name, label=Name), data=di.12_aggre, hjust=1) +
    theme_void()

f2 <- ggplot() + geom_text(aes(x=1,y=Name, label=DIrr2), data=di.12_aggre, hjust=1) +
    theme_void() +
    theme(plot.margin = unit(c(0,0,0,0), "cm"))

f3 <- ggplot() + geom_text(aes(x=1,y=Name, label=CI), data=di.12_aggre, hjust=1) +
    theme_void() +
    theme(plot.margin = unit(c(0,0,0,0), "cm"))

f4 <- ggplot(data=di.12_aggre) + geom_point(aes(x=DIrr1, y=Name)) +
    geom_vline(xintercept = 0,  linetype='dashed', col='red') +
    geom_linerange(aes(xmin=DIlr1, xmax=DIur1, y=Name)) +
    theme_void() + 
    theme(axis.line.x = element_line(), 
          axis.text.x = element_text(), 
          axis.title.x = element_text()) +
    scale_x_continuous(name='log ratio of deletion/insertion')


plot_grid(f1,f2,f3, f4, nrow=1, axis='bt', align='hv', 
          rel_widths = c(0.3,0.2,0.3,0.3))
ggsave('deletion_insertion_ratio.pdf', width = 6, height = 4)

```

# Appendix

## Del Ins Genome size

```{r}


cx <- cc[Gs_Mb!=0] %>% mutate(log_Gs_Mb = log(Gs_Mb))  %>%
    select(Species, log_Gs_Mb, Group)

dc <- di.123 %>% select(Species, DIrr) %>%
    left_join(.,cx, by="Species")


hh <- dc[!Group%in% c('human','mammals'), Group] %>% unique()
hh_c <- scol_2[names(scol_2) %in% hh]

P_wo_human_mammls <- ggplot(data=unique(dc[!Group%in% c('human','mammals')]), 
                            aes(x=log_Gs_Mb, y=DIrr)) + 
    geom_point(aes(color=Group), size=4) +
    scale_x_continuous(name='log genome size (Mb)') +
    scale_y_continuous(name='log ratio of deletion/insertion') +
    scale_color_manual(values=hh_c) +
    geom_smooth(method='lm', se=TRUE, size=1, col='darkgrey') +
    stat_cor(aes(group = 1,
                 label =paste(..rr.label.., ..p.label.., sep = "~`,`~")),
             label.x = 2,
             label.y = -1, size = 4, digits= 2, check_overlap = TRUE) +
    theme_bw()


hh_c2 <- scol_2[names(scol_2) %in% unique(dc$Group)]

P_w_human_mammls <- ggplot(data=unique(dc), aes(x=log_Gs_Mb, y=DIrr)) + 
    geom_point(aes(color=Group), size=4) +
    scale_x_continuous(name='log genome size (Mb)') +
    scale_y_continuous(name='log ratio of deletion/insertion') +
    geom_smooth(method='lm', se=TRUE, size=1, col='darkgrey') +
    scale_color_manual(values=hh_c2) +
    stat_cor(aes(group = 1,
                 label =paste(..rr.label.., ..p.label.., sep = "~`,`~")),
             label.x = 2,
             label.y = -1, size = 4, digits= 2, check_overlap = TRUE) +
    theme_bw()

plot_grid(P_wo_human_mammls, P_w_human_mammls, ncol=2, labels = c('(A)', '(B)'))


ggsave('deletion_insertion_genomeSize.png', height=6, width = 10, units = 'in', dpi = 300)

```





